cli/src/config/merge_tools.toml

```diff
@@ -11,6 +11,15 @@ merge-args = ["$left", "$base", "$right", "-o", "$output", "--auto-merge"]
 program="meld"
 # If using this as a template, note that `$output` is repeated twice below
 edit-args = ["$left", "$output", "$right", "-o", "$output"]
+# With Meld's 3-way merge resolution mode, once the output pane has been
+# edited, it loses the original contents of `$base`. Here we offer an
+# alternative that opens the original two sides compared to the base in two
+# more tabs.
+merge-args = ["--diff", "$left", "$output", "$right", "--diff", "$base", "$left", "--diff", "$base", "$right"]
```

Comment:
Have you tried making the first of these a proper merge tab (with or without auto-merge)? I'm guessing it doesn't work?

Reply:
I actually find this merge view with `jj`'s conflict presentation in the middle confusing; the way `jj` currently shows merge conflict becomes misleading in the central tab.

<img width="1849" alt="image" src="https://github.com/martinvonz/jj/assets/4123047/e9d6538e-c198-4a78-8dd1-d5db6c30fc6c">

Of the two other tabs, only one is actually useful in this case of rebase:

<img width="1849" alt="image" src="https://github.com/martinvonz/jj/assets/4123047/18529e5a-9803-47b8-8579-159509187740">

The other one is not very useful (for a rebase merge conflict, at least):

<img width="1849" alt="image" src="https://github.com/martinvonz/jj/assets/4123047/534b879d-48cf-4ffb-8809-1c78292abaa9">

There's this weird fact that even though it's clear what the files are in the two extra tabs, each tab is confusing on its own.

I think I'm still OK with using this as an "experimental" merge view, unless we come up with something better, but I'm thinking about putting more warnings in the description.

Reply:
The problem I have with Meld’s 3-pane merging mode is that, when there is a conflict (an area marked with red background), it stops highlighting the character-level changes and leaves you an opaque wall of code in which to hunt for changes, as in a Find The Differences game:

![image](https://github.com/martinvonz/jj/assets/105565/f0c4ab1b-d1f9-4980-8548-9b8af4320a72)
<sup>`meld left base right -o output --auto-merge`</sup>

This is not what I signed up for when I launched a GUI app.

If you know how to read JJ’s conflict markers, then at least you get some line-level hints with `merge-tool-edits-conflict-markers = true`. This is the first tab of `meld-3`:

![image](https://github.com/martinvonz/jj/assets/105565/b0789100-02df-4745-8f46-ebd930190cf6)
<sup>`meld left output right`</sup>

I presume I don’t need to explain the conflict markers here. Just remember that the part with `+++++++` will be identical to one of the sides, the context (` `) and removed (`-`) lines in the `%%%%%%%` part come from the base, and the added (`+`) lines would be from the side that was not chosen for `+++++++`. Older versions of Meld would have made it more obvious, as they did not have the red background for conflicts, and identical parts used to be not highlighted.

Going back to my pain point, the lack of character-level highlighting in 3-pane mode, this is what the other two tabs are for. They let you see the changes each side had made, highlighted nicely. Normally, only the “local” (right) side will be interesting, because these are the changes we would like to apply. It is perfectly normal in my workflow to ignore the middle tab, though sometimes I do like to peek into what the “other” (left) side was doing.

Side note: Using the 3-pane mode _without_ `--auto-merge` should be reserved for pathological cases where automatic merging does the [wrong](https://tahoe-lafs.org/~zooko/badmerge/simple.html) [thing](https://tahoe-lafs.org/~zooko/badmerge/concrete-good-semantics.html):

![image](https://github.com/martinvonz/jj/assets/105565/15374806-f228-44b1-b1d1-6f78395e5133)
<sup>`meld left base right -o output`</sup>

I hope I covered all of your concerns. Now I will take it unto myself to distil it down to a few succinct words to include in the TOML comment and in the docs.

Reply:
> The problem I have with Meld’s 3-pane merging mode is that, when there is a conflict (an area marked with red background), it stops highlighting the character-level changes and leaves you an opaque wall of code in which to hunt for changes, as in a Find The Differences game:

I know this problem, and it makes sense to me that showing the two extra tabs helps with it.

However, it would make more sense to me if the first tab (3-pane one) would just show the normal Meld merge view, with `--auto-merge`:

![image](https://github.com/martinvonz/jj/assets/4123047/b71c10da-d5bb-45d8-943b-430abc1a7fdf)

It seems less confusing than the view your setup shows as the first tab. I'm not sure whether Meld can show merge tabs and diff tabs at the same time, though.

So, the only difference between the `meld` and `meld-3` merge tools would be the two extra tabs.


-------------------------

Another more out-there option, especially if Meld cannot show merge tabs and diff tabs at the same time, would be to teach `jj` to produce different kinds of conflict markers. For example, I think that Git-style conflict markers would be less confusing in this UI.

Reply:
BTW, I know I said I'm thinking about putting more warnings in the description, but your text LGTM at the moment. I think you improved it a bit since I last looked (👍 ), but it was quite clear before as well. We can always improve it later if needed. It'd take me effort to try to add warnings without making it less clear, and it's quite possible that I'd conclude that it's better as is if I tried.